### PR TITLE
ref: Prototype of a generic streaming consumer strategy factory

### PR DESCRIPTION
**Pull request by lynnagara**
This is an attempt to make the streaming consumer strategy factory
independent of the Snuba specific way of writing to tables and doing
replacements so that it can eventually be used by Sentry as well.
The strategy factory only takes a generic collector function.

### CODE REVIEW

File: snuba/utils/streams/strategy_factory.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+TProcessed = TypeVar("TProcessed")
+
+
+class StreamingConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def __init__(
+        self,
+        prefilter: Optional[StreamMessageFilter],
+        process_message: Callable[[Message[KafkaPayload]], TProcessed],
+        collector: Callable[[], ProcessingStrategy[TProcessed]],
```
Comment (fpacifici) [inline snuba/utils/streams/strategy_factory.py:38]: Good thanks. I like this.

Can we parametrize the Payload instead of using `KafkaPayload` ? This abstraction seems not to be Kafka specific now.

File: snuba/utils/streams/strategy_factory.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    @abstractmethod
+    def should_drop(self, message: Message[KafkaPayload]) -> bool:
+        raise NotImplementedError
+
+
+TProcessed = TypeVar("TProcessed")
+
+
+class StreamingConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
```
Comment (fpacifici) [inline snuba/utils/streams/strategy_factory.py:33]: If this is now a reusable component I would add a proper docstring here that explains what it should be used for (consumers that do filtering and collect in batches, and can work in a parallel or serial way, etc.)?

File: snuba/consumers/consumer.py [hunk 3]
```diff
@@ -334,39 +274,7 @@ def __build_write_step(self) -> ProcessedMessageBatchWriter:
...
-                input_block_size=self.__input_block_size,
-                output_block_size=self.__output_block_size,
-                metrics=StreamMetricsAdapter(MetricsWrapper(self.__metrics, "process")),
-            )
-
-        if self.__prefilter is not None:
-            strategy = FilterStep(self.__should_accept, strategy)
-
-        return strategy
+    return build_writer
```
Comment (fpacifici) [inline snuba/consumers/consumer.py:277]: Who is actually using this function ?

File: snuba/consumers/consumer.py [hunk 4]
```diff
@@ -441,26 +349,52 @@ class MultistorageKafkaPayload(NamedTuple):
...
 
-    value = rapidjson.loads(message.payload.payload.value)
+    storages: MutableSequence[WritableTableStorage] = []
+
+    for storage_key in storage_keys:
+        storage = get_writable_storage(storage_key)
+        filter = storage.get_table_writer().get_stream_loader().get_pre_filter()
+        if filter is None or not filter.should_drop(message):
+            storages.append(storage)
+
```
Comment (fpacifici) [inline snuba/consumers/consumer.py:385]: Why this change ?

**Review by fpacifici — COMMENTED** [review]
I think this is a good approach. 
A few questions inline.

File: snuba/consumers/consumer.py (hunk 3 above)
Comment (lynnagara) [reply snuba/consumers/consumer.py:277]: This is the writer function passed to the collect step and called on every collect. It should be fundamentally the same as before where it lived in the streaming strategy factory https://github.com/getsentry/snuba/blob/master/snuba/consumers/consumer.py#L318-L335

Comment (fpacifici) [reply snuba/consumers/consumer.py:277]: Ah damn, I messed up indentation. I thought this was the return of `build_writer` itself, which would make no sense. Makes sense then.

File: snuba/consumers/consumer.py (hunk 4 above)
Comment (lynnagara) [reply snuba/consumers/consumer.py:385]: I wanted to reuse the strategy factory code here. This meant converting the multistorage strategy from `transform -> filter -> transform -> collect` into `transform -> collect`. So now this transform step includes the parts that were here before https://github.com/getsentry/snuba/blob/master/snuba/consumers/consumer.py#L617-L620

File: snuba/utils/streams/strategy_factory.py (hunk 1 above)
Comment (lynnagara) [reply snuba/utils/streams/strategy_factory.py]: I changed this class to support parametrized payload. I also added `KafkaConsumerStrategyFactory(ConsumerStrategyFactory[KafkaPayload])` though which is the one I'd expect end users to mostly use.

File: snuba/consumers/consumer.py (hunk 4 above)
Comment (fpacifici) [reply snuba/consumers/consumer.py:385]: This opens up an issue.
Before we were doing this filtering step first
```
return TransformStep(
    self.__find_destination_storages,
    FilterStep(self.__has_destination_storages, strategy,),
)
```
that was the very first thing that would happen before going to the parallel step.
For a consumer like cdc, that is multi storage, but that drops the vast majority of the messages, that meant that before we would be dropping most messages before getting into the dispatch step of the parallel transformation (the dispatch is a heavy one in the parallel consumer as it can be slowed down by lock contention when the parallel step is trivial. Most of the time would be spent in filling batches and waiting for them to be ready).
With your change this filtering is in `process_message_multistorage`, which is the parallel step and thus happens  AFTER dispatching. 

I would be concerned this would have a negative impact on the throughput of consumers like cdc.

Can we move this back to the filter ?
Also maybe leaving a note about this issue in a comment would help the next time somebody approaches this code. It can be tricky to notice at first.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Please see my comment on the filtering step on the multi storage consumer.

File: snuba/consumers/consumer.py (hunk 4 above)
Comment (lynnagara) [reply snuba/consumers/consumer.py:385]: Damn I did not realise CDC was using this strategy. Let me think about how to keep this as a filter step.

Comment (fpacifici) [reply snuba/consumers/consumer.py:385]: Since you are there. Could you please ensure you test these cases manually as we have little coverage of the consumer end to end:
1) Run a multiprocess consumer on one storage alone and ensure it consume in parallel
2) Same without switching the consumer to multi processing
3) Run a multi storage consumer with 1 process (Starting devserves now starts the cdc consumer in multi storage mode. Then just enable cdc in sentry)
4) Run a multi storage consumer with 2-3 processes

Comment (lynnagara) [reply snuba/consumers/consumer.py:387]: I've tried the 4 combinations (with the events and errors storage for multistorage consumer) and it works. Still trying to get CDC running in Sentry though.

Comment (lynnagara) [reply snuba/consumers/consumer.py:387]: Also added the multistorage filter step back here. Note we still need to re-filter in process_message_multistorage though to ensure that the event is only written to the right storage. If you think that this double filtering is too undesirable I can revert these changes and not inherit from the generic strategy factory.

Comment (fpacifici) [reply snuba/consumers/consumer.py:387]: Running the filter twice is not dangerous. Though it will be quite confusing for those who look at the code.
It seems to me this issue is actually a signal that we are trying to abstract together a few things that should not be abstracted together.
here we have serial and parallel transform on one dimension. Plus single storage and multi storage on the other dimension.
The abstraction works well for serial transformations. You can easily represent the process as filter-proess-collect. For parallel consumption this is transform-parallel_transform-collect instead.
What if we captured that difference (the two types of process) in the type parameters? Which basically requires two options:
- having a StandardConsumerStrategyFactory and a parallel one. In one the first step is a fitler in the other it is a transformation.
OR
- Turn the  ConsumerStrategyFactory into a process with three steps: pre-transform (instead of filter) transform and collect. I like the other option better since, in this case, the need for two transform step is due to an implementation detail we are trying to hide (and which cannot really be hidden).

Comment (lynnagara) [reply snuba/consumers/consumer.py]: I reverted the changes to the multistorage consumer; no more fitting this into the generic 3 steps, we're back to 4. I don't think it's worth providing something in the library to handle this particular 4 step scenario as I don't imagine
it's a common one.

Personally I thought the earlier approach was more straightforward because there wasn't the extra intermediate format that the message was transformed into to think about, but this may be highly subjective.

**Review by fpacifici — APPROVED** [review]
